ddtrace/sampler.py

```diff
@@ -276,25 +278,26 @@ def sample(self, span):
                     self._set_priority(span, AUTO_REJECT)
                     return False
 
-            # If no rules match, use our default sampler
-            matching_rule = self.default_sampler
+            # If no rules match, use our default rule sampler
+            # DEV: If it isn't a RateByServiceSampler then it must be a SamplingRule
+            matching_rule = cast(SamplingRule, self.default_sampler)
+
+        # Set a metric saying we sampled with a user defined rule
+        # DEV: Avoid `span.set_metric()` which adds constraint checking overhead
+        span.metrics[SAMPLING_RULE_DECISION] = matching_rule.sample_rate
 
         # Sample with the matching sampling rule
-        if isinstance(matching_rule, (RateSampler, SamplingRule)):
-            span.set_metric(SAMPLING_RULE_DECISION, matching_rule.sample_rate)
         if not matching_rule.sample(span):
             self._set_priority(span, USER_REJECT)
             return False
-        else:
-            # Do not return here, we need to apply rate limit
-            self._set_priority(span, USER_KEEP)
```

Comment:
Why was this branch removed?

Reply:
if you look at the end of the function we also call `self._set_priority(span, USER_KEEP)`, so this is an unnecessary extra call